### PR TITLE
Insight logic refactor version 1 - keyed `insightLogic`

### LINKED ISSUE #6105: If one dashboard item is slow to load, the rest of the (loaded) items don't load
## Bug description

* Go to a dashboard with lots of items
* Some will load fast
* If any slow loading item fails to load, it then replaces _all_ (including loaded items) with an error message

Screenshot 2021-09-26 at 14 04 11

@mariusandra perhaps related to your dashboard loading work last week?

## Expected behavior

* Only the specific dashboard items that fail to load should have the above error message

### PR DESCRIPTION
**Pull request by mariusandra**
## TL;DR

This PR tightly couples `insightLogic` and the visualisation logics (`trendsLogic`, `funnelLogic`, `pathLogic`, `retentionTableLogic`) via shared props. This might already fix https://github.com/PostHog/posthog/issues/6105

## Changes

This is the first chunk of refactors from https://github.com/PostHog/posthog/pull/6262
I really tried to scope it down, but it's hard since I have to update so many things to make this work well. A lot of the files are just a few very common changes.

 - **Main change:** Add `props` and a `key` to `insightLogic`, and make sure we always pass some props.
 - Make all instances of `trendsLogic`, `pathsLogic`, `retentionTableLogic` and `funnelLogic` also require the same props.
 - Using `<BindLogic>`, introduce a pattern to be used in any component:
```ts
const { insightProps } = useValues(insightLogic) // props set via BindLogic
const { /* whatever we need */ } = useValues(funnelLogic(insightProps))
``` 
 - Add kea test `toNotHaveDispatchedActions`, and fix some bugs on the test for `dashboardLogic`
 - Hotfix to `insightMetadataLogic` that will get a good refactor later
 - Exit edit mode and enter view mode if clicking "save" for the insight
 - Fix various bugs where different filters weren't persisting the `fromItem` insight ID in the url

## To come

This is just the first set of insight-related refactors. Still to come in the next PRs:

- Move the dozen logics away from relying on the URL. Only have one `insightsSceneLogic` that syncs things in the end.
- Create a proper flow for loading and displaying data, and editing in-flight filters (`insightLogic` will be the source of truth, not `trend/paths/.../Logic`)
- Better URL handling, where the entire query goes from `searchParams` to `hashParams`,
- Nicer URLs for saved insights like `/insights/123`
- Unified insight prop cleanup/sanitization and result fetching (vs in different sub-logics like now)
- Tighter integration between insightLogic and dashboardLogic
- Rename `dashboardItemId` -> `insightId`

## How did you test this code?

- With both the `3408-saved-insights` feature flag on and off, I tried creating, saving and exploring various insights and dashboard items. I tried all the filters, tried saving insights on dashboards, flipping between tabs, between view/edit modes, etc. Things seem to not work worse than before 😅 

## Next steps

I'd love to merge this in Friday unless someone finds things that are really broken. The bar for quality is "not worse than now". So I won't fix it if e.g. tags can't be added in view mode under saved insights, as that's a current bug. Many more logic and stability improvements will come with the next PRs on top of this.

### CODE REVIEW

**Review by mariusandra — COMMENTED** [review]
Finally tests pass and it's ready for a look. Cypress has been sporadic, but its real errors got fixed, just seems like some flakes flaring up now.

**Review by pauldambra — APPROVED** [review]
nothing in there that shouts at me to block to this... and some nice changes in amongst it...

what a beast of a PR :)

just going to pull and check it works locally (although CI is all green)

LGTM

**Comment by pauldambra** [comment]
When I'm on this branch and loading an insight view page I get 

```
VM84 react_devtools_backend.js:2526 Warning: Cannot update a component (`Title`) while rendering a different component (`ForwardRef`). To locate the bad setState() call inside `ForwardRef`, follow the stack trace as described in https://fb.me/setstate-in-render
    in ForwardRef (created by TrendTab)
    in div (created by Col)
    in Col (created by TrendTab)
    in div (created by Row)
    in Row (created by TrendTab)
    in TrendTab (created by Insights)
    in div (created by Insights)
    in div (created by Card)
    in div (created by Card)
    in Card (created by Insights)
    in div (created by Col)
    in Col (created by Insights)
    in div (created by Row)
    in Row (created by Insights)
    in div (created by Insights)
    in BindLogic (created by Insights)
    in Insights (created by AppScene)
    in main (created by Basic)
    in Basic (created by Content)
    in Content (created by AppScene)
    in section (created by BasicLayout)
    in BasicLayout (created by Layout)
    in Layout (created by AppScene)
    in section (created by BasicLayout)
    in BasicLayout (created by Layout)
    in Layout (created by AppScene)
    in AppScene (created by App)
    in App
    in Provider
```

but... on current HEAD I get

```
Warning: Cannot update a component (`CommandPalette`) while rendering a different component (`Insights`). To locate the bad setState() call inside `Insights`, follow the stack trace as described in https://fb.me/setstate-in-render
    in Insights (created by AppScene)
    in main (created by Basic)
    in Basic (created by Content)
    in Content (created by AppScene)
    in section (created by BasicLayout)
    in BasicLayout (created by Layout)
    in Layout (created by AppScene)
    in section (created by BasicLayout)
    in BasicLayout (created by Layout)
    in Layout (created by AppScene)
    in AppScene (created by App)
    in App
    in Provider
```

So I guess it's an existing non-breaking error that has moved with the changes??

**Comment by mariusandra** [comment]
Thanks for the review! Fixed a few small things, but will push others to [the next PRs](https://github.com/PostHog/posthog/pull/6327). 

I'm not sure where these invalid setState calls come from, though they seem quite inconsequential. They also only show up when you have the saved reports flag enabled, and I really want to redo the `Title` component anyway, so I'll fix this after this PR as well.
